File: machine-learning-ex4/r/ex4.r
## Machine Learning Online Class (R implementation) - Exercise 4 Neural Network Learning
require('rdetools')
require('R.matlab')
rm(list = ls())  # Clear environment variables
lapply(list.files(pattern = "[.]R$", recursive = T), source)  # load all fn's

input_layer_size <- 400  # 20x20 input images of digits
hidden_layer_size <- 25  # 25 hidden units
num_labels <- 10         # 10 labels, from 1 to 10 ("0" mapped to label 10)

## =========== Part 1: Loading and Visualizing Data =============
cat("\nLoading and visualizing data...\n")

# Load training data
data <- readMat('ex4data1.mat')
X <- data$X
y <- data$y
m <- nrow(X)  # number of training examples

# Randomly select 100 data points to display
rand_indices <- sample(m, 100)
sel <- X[rand_indices,]

displayData(sel)

readline(prompt = "Program paused. Press enter to continue.")

## ================ Part 2: Loading Parameters ================
cat("\nLoading saved neural network parameters...\n")
params <- readMat('ex4weights.mat')

theta1 <- params$Theta1
theta2 <- params$Theta2

# Unroll parameters
nn_params <- rbind(matrix(theta1, ncol = 1), matrix(theta2, ncol = 1))

## ================ Part 3: Compute Cost (Feedforward) ================
cat("\nFeedforward using neural network...\n")

# Weight regularization parameter (set to 0 here)
lambda <- 0
cg <- nnCostFunction(nn_params, input_layer_size, hidden_layer_size,
                    num_labels, X, y, lambda)
cost <- cg$cost
grad <- cg$grad

cat("Cost at parameters:", cost, "\n")
cat("  (this value should be about 0.287629)\n")

readline(prompt = "Program paused. Press enter to continue.")

## =============== Part 4: Implement Regularization ===============
cat("\nChecking cost function (w/ regularization)...\n")

# Weight regularization parameter (now we set to 1 here)
lambda <- 1

cg <- nnCostFunction(nn_params, input_layer_size, hidden_layer_size,
                    num_labels, X, y, lambda)
cost <- cg$cost
grad <- cg$grad

cat("Cost at parameters:", cost, "\n")
cat("  (this value should be about 0.383770)\n")

readline(prompt = "Program paused. Press enter to continue.")

## ================ Part 5: Sigmoid Gradient  ================
cat("\nEvaluating sigmoid gradient...\n")

g <- sigmoidGradient(c(-1, -0.5, 0, 0.5, 1))
cat("Sigmoid gradient evaluated at c(-1, -0.5, 0, 0.5, 1):\n")
cat(" ", g, "\n")
cat("(these values should be:)\n")
cat("  0.196612 0.235004 0.25 0.235004 0.196612")

readline(prompt = "Program paused. Press enter to continue.")

## ================ Part 6: Initializing Pameters ================
cat("\nInitializing neural network parameters...\n")

initialTheta1 <- randInitializeWeights(input_layer_size, hidden_layer_size)
initialTheta2 <- randInitializeWeights(hidden_layer_size, num_labels)

# Unroll parameters
initial_nn_params <- rbind(matrix(initialTheta1, ncol = 1),
                          matrix(initialTheta2, ncol = 1))

## =============== Part 7: Implement Backpropagation ===============
cat("\nChecking backpropagation...\n")
checkNNGradients()

readline(prompt = "Program paused. Press enter to continue.")

## =============== Part 8: Implement Regularization ===============
cat("\nChecking backpropagation (w/ regularization)...\n")

# Check gradients by running checkNNGradients
lambda <- 3
checkNNGradients(lambda)

# Also output the costFunction debugging values
debug_cg <- nnCostFunction(nn_params, input_layer_size, hidden_layer_size,
                          num_labels, X, y, lambda)$cost
cat("Cost at (fixed) debugging parameters (w/ lambda =", lambda, "):", debug_cg, "\n")
cat("  (for lambda = 3, this value should be about 0.576051)\n")

readline(prompt = "Program paused. Press enter to continue.")

## =================== Part 9: Training NN ===================
cat("\nTraining neural network...\n")

lambda <- 1

# Optimize using fmincg:
cat("\n  Training via fmincg...\n")
costFn = function(p) nnCostFunction(p, input_layer_size, hidden_layer_size,
                                    num_labels, X, y, lambda, F)
options = c(MaxIter = 250)
fcg = fmincg(costFn, initial_nn_params, options)
nn_params_fmincg = fcg$params
cost_fmincg = fcg$cost

# Optimize using optim()
cat("\n  Training via optim...\n")
getCostGrad = function(p, getNew) {
  # We'll save the output from nnCostFunction to global variable getCG, then use
  # getNew to denote whether or not we want to run nnCostFunction again. This
  # will only happen when we want to get a new cost, at which point we'll get a
  # new parameter gradient, too.
  if (getNew) {
    assign("getCG", nnCostFunction(p, input_layer_size, hidden_layer_size,
                                   num_labels, X, y, lambda, T),
           envir = .GlobalEnv)
    return(getCG$cost)
  } else {
    return(getCG$grad)
  }
}

getCost <- function(p) getCostGrad(p, T)
getGrad <- function(p) getCostGrad(p, F)

# optim() doesn't have a TNC optimization method
res <- optim(initial_nn_params, fn = getCost, gr = getGrad,
             method = 'BFGS', control = list(maxit = 250))
nn_params_optim <- res$par
cost_optim <- res$value

cat("\nCost via fmincg:", cost_fmincg[length(cost_fmincg)])
cat("\nCost via optim: ", cost_optim, "\n")

if (cost_fmincg[length(cost_fmincg)] < cost_optim) {
  cost = cost_fmincg[length(cost_fmincg)]
  nn_params = nn_params_fmincg
} else {
  cost = cost_optim
  nn_params = nn_params_optim
}

# Obtain theta1 and theta2 back from nn_params
paramunroll <- hidden_layer_size * (input_layer_size + 1)
theta1 <- matrix(nn_params[1:paramunroll],
                nrow = hidden_layer_size, ncol = input_layer_size + 1)
theta2 <- matrix(nn_params[(paramunroll + 1):length(nn_params)],
                nrow = num_labels, ncol = hidden_layer_size + 1)

readline(prompt = "Program paused. Press enter to continue.")

## ================= Part 10: Visualize Weights =================
cat("\nVisualizing neural network...\n")

displayData(theta1[,-1])

readline(prompt = "Program paused. Press enter to continue.")

## ================= Part 11: Implement Predict =================
pred <- predict(theta1, theta2, X)

cat("\nTraining set accuracy:", mean(pred == y) * 100, "%\n")

File: machine-learning-ex4/python/ex4.py
## Machine Learning Online Class (Python implementation) - Exercise 4 Neural Network Learning
import matplotlib.pyplot as plt
import numpy as np
import scipy.io as sio
import scipy.optimize as op
from ex4lib import *

if __name__ == '__main__':
    input_layer_size = 400  # 20x20 input images of digits
    hidden_layer_size = 25  # 25 hidden units
    num_labels = 10         # 10 labels, from 1 to 10 ("0" mapped to label 10)

## =========== Part 1: Loading and Visualizing Data =============
    print("\nLoading and visualizing data...")
    data = sio.loadmat('ex4data1')
    X = data['X']
    y = data['y']
    m, n = X.shape

    # Randomly select 100 data points to display
    rand_indices = np.random.choice(m, size = 100, replace = False)
    sel = X[rand_indices, :]

    displayData(sel)
    plt.show()

    input("Program paused. Press enter to continue.")

## ================ Part 2: Loading Parameters ================
    print("\nLoading saved neural network parameters...")
    params = sio.loadmat('ex4weights')
    theta1 = params['Theta1']
    theta2 = params['Theta2']

    # Going to ignore the parameter unrolling here because it isn't necessary
    # in Python.
    nn_params = np.append(theta1.flatten(), theta2.flatten())
## ================ Part 3: Compute Cost (Feedforward) ================
    print("\nFeedforward using neural network...")

    # Weight regularization parameter (we set this to 0 here)
    lamb = 0

    J, grad = nnCostFunction(nn_params, input_layer_size, hidden_layer_size,
                             num_labels, X, y, lamb)
    
    print("Cost at parameters: %.6f" % J)
    print("  (this value should be about 0.287629)")

    input("Program paused. Press enter to continue.")

## =============== Part 4: Implement Regularization ===============
    print("\nChecking cost function (w/ regularization)...")

    # Weight regularization parameter (we now set this to 1 here)
    lamb = 1

    J, grad = nnCostFunction(nn_params, input_layer_size, hidden_layer_size,
                             num_labels, X, y, lamb)
                            
    print("Cost at parameters: %.6f" % J)
    print("  (this value should be about 0.383770)")

    input("Program paused. Press enter to continue.")

## ================ Part 5: Sigmoid Gradient  ================
    print("\nEvaluating sigmoid gradient...")

    g = sigmoidGradient(np.array([-1, -0.5, 0, 0.5, 1]))
    print("Sigmoid gradient evaluated at [-1, -0.5, 0, 0.5, 1]:")
    print("\t", g)
    print("  (these values should be:)")
    print("\t [ 0.196612\t0.235004\t0.250000\t0.235004\t0.196612]")

    input("Program paused. Press enter to continue.")

## ================ Part 6: Initializing Pameters ================
    print("\nInitializing neural network parameters...")

    initialTheta1 = randInitialWeights(input_layer_size, hidden_layer_size)
    initialTheta2 = randInitialWeights(hidden_layer_size, num_labels)

    initial_nn_params = np.append(initialTheta1.flatten(), initialTheta2.flatten())

## =============== Part 7: Implement Backpropagation ===============
    print("\nChecking backpropagation...")

    # Check gradients by running checkNNGradients
    checkNNGradients()

## =============== Part 8: Implement Regularization ===============
    print("\nChecking backpropagation (w/ regularization)")

    # Check gradients by running checkNNGradients
    lamb = 3
    checkNNGradients(lamb)

    # Also output the costFunction debugging values
    debug_J = nnCostFunction(nn_params, input_layer_size, hidden_layer_size,
                             num_labels, X, y, lamb)[0]
            
    print("Cost at (fixed) debugging parameters (w/ lambda = %.3f): %.3f" % (lamb, debug_J))
    print("  (for lambda = 3, this value should be about 0.576051)")

    input("Program paused. Press enter to continue.")

## =================== Part 9: Training NN ===================
    print("\nTraining neural network... ")

    lamb = 1
    costFn = lambda p: nnCostFunction(p, input_layer_size, hidden_layer_size,
                                      num_labels, X, y, lamb, False)

    # op.minimize() requires/works best when x0 is a single vector of type ndarray
    # if jac = True, costFn is assumed to return the cost as the first return
    # value, and the gradient as the second.
    res = op.minimize(fun = costFn, x0 = initial_nn_params, jac = True,
                        method = 'TNC', options = {'maxiter': 250})
    cost = res.fun
    nn_params = res.x

    # Obtain theta1 and theta2 back from nn_params
    paramunroll = hidden_layer_size * (input_layer_size + 1)
    theta1 = np.reshape(nn_params[:paramunroll], [hidden_layer_size, input_layer_size + 1])
    theta2 = np.reshape(nn_params[paramunroll:], [num_labels, hidden_layer_size + 1])

    input("Program paused. Press enter to continue.")

## ================= Part 10: Visualize Weights =================
    # Visualize what the neural network is learning by displaying the hidden
    # units to see what featuers they are capturing in the data.
    print("\nVisualizing neural network...")
    displayData(theta1[:, 1:])
    plt.show()
    
    input("Program paused. Press enter to continue.")

## ================= Part 11: Implement Predict =================
    # Use the neural network to predict the labels
    pred = predict(theta1, theta2, X)
    ym = np.reshape(y, pred.shape)

    print("\nTraining set accuracy: %.3f%%" % (np.mean(pred == ym) * 100))
